osu.Game/Utils/ModUtils.cs

```diff
@@ -0,0 +1,132 @@
+            => CheckCompatibleSet(combination, out _);
+
+        /// <summary>
+        /// Checks that all <see cref="Mod"/>s in a combination are compatible with each-other.
+        /// </summary>
+        /// <param name="combination">The <see cref="Mod"/> combination to check.</param>
+        /// <param name="invalidMods">Any invalid mods in the set.</param>
+        /// <returns>Whether all <see cref="Mod"/>s in the combination are compatible with each-other.</returns>
+        public static bool CheckCompatibleSet(IEnumerable<Mod> combination, out List<Mod>? invalidMods)
+        {
+            combination = FlattenMods(combination).ToArray();
+            invalidMods = null;
+
+            foreach (var mod in combination)
+            {
+                foreach (var type in mod.IncompatibleMods)
+                {
+                    foreach (var invalid in combination.Where(m => type.IsInstanceOfType(m)))
+                    {
+                        invalidMods ??= new List<Mod>();
+                        invalidMods.Add(invalid);
+                    }
+                }
+            }
+
+            return invalidMods == null;
+        }
+
+        /// <summary>
+        /// Checks that all <see cref="Mod"/>s in a combination appear within a set of allowed types.
+        /// </summary>
+        /// <remarks>
+        /// The set of allowed types must contain exact <see cref="Mod"/> types for the respective <see cref="Mod"/>s to be allowed.
+        /// </remarks>
+        /// <param name="combination">The <see cref="Mod"/> combination to check.</param>
+        /// <param name="allowedTypes">The set of allowed <see cref="Mod"/> types.</param>
+        /// <returns>Whether all <see cref="Mod"/>s in the combination are allowed.</returns>
+        public static bool CheckAllowed(IEnumerable<Mod> combination, IEnumerable<Type> allowedTypes)
+        {
+            var allowedSet = new HashSet<Type>(allowedTypes);
+
+            return combination.SelectMany(FlattenMod)
+                              .All(m => allowedSet.Contains(m.GetType()));
+        }
+
+        /// <summary>
+        /// Check the provided combination of mods are valid for a local gameplay session.
+        /// </summary>
+        /// <param name="mods">The mods to check.</param>
+        /// <param name="invalidMods">Invalid mods, if any were found. Can be null if all mods were valid.</param>
+        /// <returns>Whether the input mods were all valid. If false, <paramref name="invalidMods"/> will contain all invalid entries.</returns>
+        public static bool CheckValidForGameplay(IEnumerable<Mod> mods, out List<Mod>? invalidMods)
+        {
+            mods = mods.ToArray();
+
+            CheckCompatibleSet(mods, out invalidMods);
+
+            foreach (var mod in mods)
+            {
+                if (mod.Type == ModType.System || !mod.HasImplementation || mod is MultiMod)
```

Comment:
Do we need the `MultiMod` check any more? You are handling the flattening of these in `CheckCompatibleSet` method, which suggests they should maybe be supported as inputs? Or is it never the case that they will be provided for a gameplay scenario?

Reply:
Actually it technically is kind-of an unsupported input for gameplay purposes... Like if you were to give `Player` a `MultiMod`, neither it nor any ruleset components would know how to deal with it.

That was my reasoning as to why I left it.